ultra/baselines/common/baseline_state_preprocessor.py

```diff
@@ -0,0 +1,193 @@
+from ultra.baselines.common.state_preprocessor import StatePreprocessor
+from ultra.utils.common import (
+    get_closest_waypoint,
+    rotate2d_vector,
+    normalize_im,
+)
+
+
+class BaselineStatePreprocessor(StatePreprocessor):
+    """The State Preprocessor used by the baseline agents."""
+    _NORMALIZATION_REFERENCES = {
+        "speed": 30.0,
+        "distance_from_center": 1.0,
+        "steering": 3.14,  # radians
+        "angle_error": 3.14,  # radians
+        "relative_goal_position": 100.0,
+        "action": 1.0,  # 2
+        "waypoints_lookahead": 10.0,
+        "road_speed": 30.0,
+    }
+
+    def __init__(
+        self,
+        social_vehicle_config,
+        observation_waypoints_lookahead,
+        action_size,
+    ):
+        self._state_description = {
+            "images": {},
+            "low_dim_states": {
+                "speed": 1,
+                "distance_from_center": 1,
+                "steering": 1,
+                "angle_error": 1,
+                "relative_goal_position": 2,
+                "action": int(action_size),  # 2
+                "waypoints_lookahead": 2 * int(observation_waypoints_lookahead),
+                "road_speed": 1,
+            },
+            "social_vehicles": int(social_vehicle_config["num_social_features"])
+            if int(social_vehicle_config["social_capacity"]) > 0
+            else 0,
+        }
+
+    @property
+    def num_low_dim_states(self):
+        return sum(self._state_description["low_dim_states"].values())
+
+    def _preprocess_state(
+        self,
+        state,
+        observation_num_lookahead,
+        social_capacity,
+        social_vehicle_config,
+        prev_action,
+        normalize=False,
+        unsqueeze=False,
+        device=None,
+    ):
+        # Obtain the information from the state needed by the baselines.
+        state = BaselineStatePreprocessor._adapt_observation_for_baseline(state)
+
+        # Get the images from the state.
+        images = {}
+        for k in self._state_description["images"]:
+            image = torch.from_numpy(state[k])
+            image = image.unsqueeze(0) if unsqueeze else image
+            image = image.to(device) if device else image
+            image = normalize_im(image) if normalize else image
+            images[k] = image
+
+        # Set the action in the basic state as the previous action.
+        state["action"] = prev_action
+
+        # Obtain the next waypoints.
+        _, lookahead_waypoints = get_closest_waypoint(
+            num_lookahead=observation_num_lookahead,
+            goal_path=state["goal_path"],
+            ego_position=state["ego_position"],
+            ego_heading=state["heading"],
+        )
+        state["waypoints_lookahead"] = np.hstack(lookahead_waypoints)
+
+        # Normalize the low-dimensional states and concatenate them.
+        normalized = [
+            BaselineStatePreprocessor._normalize(key, state[key])
```

Comment:
Oh yeah for sure, the static methods in BaselineStatePreprocessor and StatePreprocessor are static because they don't require any instance attributes to perform the method. What are your thoughts on this?

Additionally, yes I agree it should be self._normalize(...) and also self.extract_closest_waypoint(...) in the next comment. Then should other calls to static methods like in _adapt_observation_for_baseline and _normalize be replaced with self.<method_name>? I guess this would mean that _adapt_observation_for_baseline and _normalize would no longer be static methods.

Reply:
the normalize method is just used for baselines, other people may need to define their own parameters for normalization I think it doesn't need to be static, is there anywhere in the code that calls the normalization? if not it's safe to keep it a method inside baseline_state_preprocessor. The same idea for adapt module, if others decide to use baseline they're gonna use the entire preprocessor not just one of the methods (at least I'm not foreseeing it)

Reply:
Ahh, I see what you mean. Sure, I will make them both instance methods.